### PR TITLE
Convert Linker to TypeScript

### PR DESCRIPTION
**Pull request by mofojed**
- Convert the Linker to TypeScript
- Precursor to fully separating dashboard out #23

### CODE REVIEW

File: packages/code-studio/src/dashboard/linker/Linker.tsx [hunk 1]
```diff
@@ -248,21 +321,21 @@ class Linker extends Component {
...
         return;
       }
 
-      // FILTER_SOURCE links have a limit of 1 link per target
+      // filterSource links have a limit of 1 link per target
       // New link validation passed, delete existing links before adding the new one
-      if (type === LinkType.FILTER_SOURCE) {
+      if (type === 'filterSource') {
         const { links } = this.props;
         const existingLinkPanelId = isReversed ? start.panelId : end.panelId;
```
Comment (mattrunyon) [inline packages/code-studio/src/dashboard/linker/Linker.tsx:333]: This might be safer in a switch/case statement since TS will check it exhaustively. If we were to add a new link state for any reason, in theory it should immediately catch that it's unhandled here (or caught by a default)

File: packages/code-studio/src/dashboard/linker/Linker.tsx [hunk 2]
```diff
@@ -152,7 +222,7 @@ class Linker extends Component {
       link => link.start?.panelId === panelId || link.end?.panelId === panelId
```
Comment (mattrunyon) [inline packages/code-studio/src/dashboard/linker/Linker.tsx:222]: My VSCode is yelling at me about this being implicitly any. I think it's something to do with the `ConnectedProps` in `LinkerProps` because when I remove that this (and 4 or 5 other spots) have proper types instead of any

We don't have a linter rule and I'm not sure that one exists for it. I found some that are close. They require a bit of config w/ eslint since you have to tell it about the TS runtime so it can do type analysis

File: packages/code-studio/src/dashboard/linker/Linker.tsx [hunk 3]
```diff
@@ -29,15 +29,85 @@ import {
...
+  timeZone: string;
+}
+
+interface OwnProps {
+  layout: GoldenLayout;
+  panelManager: PanelManager;
+  localDashboardId: string;
+}
+
+const mapState = (state: LinkerState, ownProps: OwnProps) => ({
```
Comment (mattrunyon) [inline packages/code-studio/src/dashboard/linker/Linker.tsx:82]: This needs an explicit return type currently since it's inferring from JS as any. So it has the right structure, but every value is any. I believe it's related to the `ConnectedProps` problem I mentioned further down in this file

File: packages/code-studio/src/dashboard/linker/Linker.tsx [hunk 4]
```diff
@@ -29,15 +29,85 @@ import {
...
+  activeTool: getActiveTool(state),
+  isolatedLinkerPanelId: getIsolatedLinkerPanelIdForDashboard(
+    state,
+    ownProps.localDashboardId
+  ),
+  links: getLinksForDashboard(state, ownProps.localDashboardId),
+  timeZone: getTimeZone(state),
+});
+
+const connector = connect(mapState, {
```
Comment (mattrunyon) [inline packages/code-studio/src/dashboard/linker/Linker.tsx:92]: Also probably needs explicit return type (or actions moved to TS). This one may be fixed by `mapState` having an explicit return type. Or you can specify `connect<State>(...)` possibly

File: packages/code-studio/src/dashboard/linker/LinkerLink.tsx [hunk 5]
```diff
@@ -15,28 +14,52 @@ const TRIANGLE_HYPOTENUSE = Math.sqrt(
...
-export class LinkerLink extends PureComponent {
-  static makeCirclePath(x, y, r) {
+export type LinkerLinkProps = {
+  x1: number;
+  y1: number;
+  x2: number;
+  y2: number;
+  id: string;
+  className: string;
+  onClick: (id: string) => void;
```
Comment (mattrunyon) [inline packages/code-studio/src/dashboard/linker/LinkerLink.tsx:24]: Since the param names in function types are just for documentation, you might be able to make the param name more descriptive? `clickedId` or `thisId` or `linkId`? Or leave it. Doesn't really matter

File: packages/code-studio/src/dashboard/linker/LinkerOverlayContent.test.tsx [hunk 6]
```diff
@@ -6,7 +6,7 @@ import LinkerOverlayContent from './LinkerOverlayContent';
 const LINKER_OVERLAY_MESSAGE = 'TEST_MESSAGE';
 
 function makeEventHub() {
-  const callbacks = {};
+  const callbacks = {} as Record<string, (arg: unknown) => void>;
```
Comment (mattrunyon) [inline packages/code-studio/src/dashboard/linker/LinkerOverlayContent.test.tsx:9]: I think you can move the type before the equals. `const callbacks: Record<> = {};`. Not sure if there's any difference between the 2

File: packages/code-studio/src/dashboard/linker/LinkerOverlayContent.tsx [hunk 7]
```diff
@@ -1,45 +1,78 @@
...
   }
 
-  componentDidMount() {
+  componentDidMount(): void {
     window.addEventListener('mousemove', this.handleMouseMove, true);
   }
 
   // eslint-disable-next-line react/sort-comp
-  componentDidCatch(error, info) {
+  componentDidCatch?(error: Error, info: ErrorInfo): void {
```
Comment (mattrunyon) [inline packages/code-studio/src/dashboard/linker/LinkerOverlayContent.tsx:66]: What is the question mark for?

File: packages/code-studio/src/dashboard/linker/LinkerOverlayContent.tsx [hunk 8]
```diff
@@ -121,15 +154,14 @@ export class LinkerOverlayContent extends Component {
...
-      .filter(item => item != null);
+      .filter(item => item != null) as VisibleLink[];
 
     return (
       <div className="linker-overlay">
         <svg>
           {visibleLinks.map(({ x1, y1, x2, y2, id, className }) => (
             <LinkerLink
               className={className}
-              disabled={disabled}
```
Comment (mattrunyon) [inline packages/code-studio/src/dashboard/linker/LinkerOverlayContent.tsx:132]: Why was this deleted with no replacement?

File: packages/iris-grid/src/formatters/TableColumnFormatter.js [hunk 9]
```diff
@@ -39,7 +39,7 @@ class TableColumnFormatter {
 
   /**
    * @param {Object} value The value to format
-   * @param {Object} format Optional format object with value transformation options
+   * @param {number} format Optional format object with value transformation options
```
Comment (mattrunyon) [inline packages/iris-grid/src/formatters/TableColumnFormatter.js:42]: The changed type and comment don't align. Type is number, comment says object

File: packages/code-studio/src/dashboard/linker/LinkerUtils.tsx [hunk 10]
```diff
@@ -3,10 +3,37 @@ import { TableUtils } from '@deephaven/iris-grid';
...
 
+export type LinkPointUntyped = {
+  panelId: string;
+  panelComponent?: string;
+  columnName: string;
+};
+
+export type LinkPoint = LinkPointUntyped & {
+  columnType: string;
+};
```
Comment (mattrunyon) [inline packages/code-studio/src/dashboard/linker/LinkerUtils.tsx:17]: Idk if changing these to `LinkPoint` and `TypedLinkPoint` would be better. Not sure about naming conventions really. Or `LinkPointWityType`.

I also only see `LinkPointUntyped` used in 1 place and I'm not sure why these couldn't just be a single type. If you wanted a partial type, you could use `Pick<LinkPoint, 'panelId' | 'columnName'>` in that spot instead if you only want to enforce it has those 2 props and they're always the same type as `LinkPoint` here.

File: packages/code-studio/src/dashboard/linker/LinkerUtils.tsx [hunk 11]
```diff
@@ -98,13 +139,14 @@ class LinkerUtils {
    * @param {string} cloneId Cloned panel id
    * @returns {object[]} Cloned links array or empty array if no new links added
```
Comment (mattrunyon) [inline packages/code-studio/src/dashboard/linker/LinkerUtils.tsx:140]: Remove types from comment

File: packages/code-studio/src/dashboard/linker/LinkerUtils.tsx [hunk 12]
```diff
@@ -98,13 +139,14 @@ class LinkerUtils {
...
    */
-  static cloneLinksForPanel(links, panelId, cloneId) {
-    const clonedLinks = [];
+  static cloneLinksForPanel(
+    links: Link[],
+    panelId: string,
+    cloneId: string
+  ): Link[] {
+    const clonedLinks = [] as Link[];
```
Comment (mattrunyon) [inline packages/code-studio/src/dashboard/linker/LinkerUtils.tsx:147]: Could use `const clonedLinks: Link[] = [];`. Again not aware of any difference for either syntax

**Review by mattrunyon — CHANGES_REQUESTED** [review]

File: packages/code-studio/src/dashboard/linker/LinkerOverlayContent.tsx (hunk 8 above)
Comment (mofojed) [reply packages/code-studio/src/dashboard/linker/LinkerOverlayContent.tsx:132]: The prop didn't exist on LinkerLink at all. It just passes a `.disabled` class on the className now.

File: packages/code-studio/src/dashboard/linker/Linker.tsx [hunk 13]
```diff
@@ -1,7 +1,7 @@
-import React, { Component } from 'react';
-import { connect } from 'react-redux';
-import PropTypes from 'prop-types';
+import React, { Component, ErrorInfo } from 'react';
+import { connect, ConnectedProps } from 'react-redux';
 import shortid from 'shortid';
+import GoldenLayout from 'golden-layout';
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/code-studio/src/dashboard/linker/Linker.tsx:4]: Nice

File: packages/code-studio/src/dashboard/linker/Linker.tsx [hunk 14]
```diff
@@ -29,15 +29,85 @@ import {
...
+  glEventHub: GoldenLayout.EventEmitter;
+};
+
+export type Panel = Component<PanelProps>;
+
+export type LinkFilterMapValue = {
+  columnType: string;
+  text: string;
+  value: unknown;
+};
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/code-studio/src/dashboard/linker/Linker.tsx:48]: Yea good call, I'll add the template to default to T = unknown

File: packages/code-studio/src/dashboard/linker/Linker.tsx [hunk 15]
```diff
@@ -29,15 +29,85 @@ import {
...
+
+export type LinkFilterMap = Map<string, LinkFilterMapValue>;
+
+export type LinkDataMapValue = {
+  type: string;
+  text: string;
+  value: string;
+};
+
+export type LinkDataMap = Record<string, LinkDataMapValue>;
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/code-studio/src/dashboard/linker/Linker.tsx:58]: It's the way it was. Not changing any of the APIs right now

File: packages/code-studio/src/dashboard/linker/Linker.tsx [hunk 16]
```diff
@@ -29,15 +29,85 @@ import {
...
+
+export type LinkablePanel = Panel & {
+  setFilterMap: (filterMap: LinkFilterMap) => void;
+  unsetFilterValue: () => void;
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/code-studio/src/dashboard/linker/Linker.tsx:62]: Yea true. I think I prefer the arrow style, since it's the similar syntax as if you're doing any other type (can mark readonly, make it optional, etc), and I think you'd have to do it like that anyway if you were doing a union or anything like that (such as `ContextAction[] | (() => ContextAction[])`. 
I don't really know the advantages the other way - you save a couple chars, and when you hover over it, it shows up as `method` instead of `property`... there's probably something I'm missing.

File: packages/code-studio/src/dashboard/linker/LinkerUtils.tsx (hunk 10 above)
Comment (mofojed) [reply packages/code-studio/src/dashboard/linker/LinkerUtils.tsx:17]: Yea I just got rid of LinkPointUntyped - I don't think it was actually needed.

File: packages/code-studio/src/dashboard/linker/LinkerUtils.tsx (hunk 12 above)
Comment (mofojed) [reply packages/code-studio/src/dashboard/linker/LinkerUtils.tsx:147]: Yea I don't care too much one way or the other. I guess having the type explicitly after the variable name is clearer than inferring the type with an `as` later

File: packages/code-studio/src/dashboard/linker/Linker.tsx [hunk 17]
```diff
@@ -76,16 +146,16 @@ class Linker extends Component {
     }
   }
 
-  componentDidCatch(error, info) {
+  componentDidCatch?(error: Error, info: ErrorInfo): void {
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/code-studio/src/dashboard/linker/Linker.tsx:153]: ??? I don't know how that got there

File: packages/code-studio/src/dashboard/linker/Linker.tsx [hunk 18]
```diff
@@ -184,49 +254,52 @@ class Linker extends Component {
...
    */
   columnSelected(
-    panel,
-    column,
+    panel: Panel,
+    column: LinkColumn,
     isAlwaysEndPoint = false,
-    overrideIsolatedLinkerPanelId = null
+    overrideIsolatedLinkerPanelId?: string
   ) {
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/code-studio/src/dashboard/linker/Linker.tsx:284]: It did not. I should probably look into _why_ it's not showing up in the linter. I've added the return type.

File: packages/code-studio/src/dashboard/linker/Linker.tsx (hunk 3 above)
Comment (mofojed) [reply packages/code-studio/src/dashboard/linker/Linker.tsx:86]: Yup adding the return type here fixes the `any` problem.

File: packages/code-studio/src/dashboard/linker/LinkerLink.tsx (hunk 5 above)
Comment (mofojed) [reply packages/code-studio/src/dashboard/linker/LinkerLink.tsx:24]: I'm going to leave it

File: packages/code-studio/src/dashboard/linker/LinkerOverlayContent.tsx (hunk 7 above)
Comment (mofojed) [reply packages/code-studio/src/dashboard/linker/LinkerOverlayContent.tsx:66]: I think I copied from the d.ts and it had it there. Fixed

File: packages/code-studio/src/dashboard/linker/Linker.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (mattrunyon) [reply packages/code-studio/src/dashboard/linker/Linker.tsx:62]: https://stackoverflow.com/questions/39156315/function-property-vs-method

Looks like you can apply readonly to a property, but not a method. Also there are some more strict checks if generics are used for a property vs if it is a method.

Linked in that thread is this page which has an example https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-6.html#example

So it looks like as a property might be the slightly safer type

[the remark this answers is not in the document]
Comment (mofojed) [reply packages/code-studio/src/dashboard/linker/Linker.tsx:62]: I'll stick with property for now

File: packages/code-studio/src/dashboard/linker/Linker.tsx [hunk 19]
```diff
@@ -41,10 +41,10 @@ export type PanelProps = {
...
 
-export type LinkFilterMapValue = {
+export type LinkFilterMapValue<T = unknown> = {
   columnType: string;
   text: string;
-  value: unknown;
+  value: T;
 };
 
 export type LinkFilterMap = Map<string, LinkFilterMapValue>;
```
Comment (mattrunyon) [inline packages/code-studio/src/dashboard/linker/Linker.tsx:50]: I think with the generic we might want to add it here as well with a default.
```suggestion
export type LinkFilterMap<T = unknown> = Map<string, LinkFilterMapValue<T>>;
```

File: packages/iris-grid/src/formatters/DateTimeColumnFormatter.ts [hunk 20]
```diff
@@ -1,17 +1,19 @@
 /* eslint class-methods-use-this: "off" */
 import dh from '@deephaven/jsapi-shim';
 import Log from '@deephaven/log';
-import TableColumnFormatter from './TableColumnFormatter';
+import TableColumnFormatter, {
+  TableColumnFormat,
```
Comment (mattrunyon) [inline packages/iris-grid/src/formatters/DateTimeColumnFormatter.ts:5]: Just FYI I looked into the `import type` more and what it does is enforces that something that isn't a type can only be used for its types within the file.

If you use a regular import for a class and then only use it for its types, it will be stripped from the compiled JS import. So I guess import type probably isn't needed other than to be explicit. Pretty sure that previous comment I had about `import type` would result in the same code regardless of the import type.

Basically import type should help if there's a build failure due to the isolated modules flag and other than that it's just us being explicit. Isolated modules flag will cause an error if there's some imports that. I think this example is an example of that flag causing a failure: https://github.com/microsoft/TypeScript/issues/28481#issuecomment-441284387

Basically re-exporting types is problematic for Babel because it only looks at a single file. So it has no idea if the re-exported value is a type or a value (I think it errs on considering it a value). That can result in unnecessary/broken imports in the final code

https://github.com/microsoft/TypeScript/pull/35200#issuecomment-645565073

File: packages/iris-grid/src/formatters/TableColumnFormatter.ts [hunk 21]
```diff
@@ -0,0 +1,69 @@
+/* eslint class-methods-use-this: "off" */
+/* eslint no-unused-vars: "off" */
```
Comment (mattrunyon) [inline packages/iris-grid/src/formatters/TableColumnFormatter.ts:2]: Curious why this rule is ignore. Can't think of any reason unused vars would be needed in a file

File: packages/iris-grid/src/formatters/TableColumnFormatter.ts [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+  static TYPE_CONTEXT_CUSTOM: TableColumnFormatType = 'type-context-custom';
+
+  /**
+   * Validates format object
+   * @param format Format object
+   * @returns true for valid object
+   */
+  static isValid(format: TableColumnFormat): boolean {
+    return true;
+  }
```
Comment (mattrunyon) [inline packages/iris-grid/src/formatters/TableColumnFormatter.ts:33]: Can you add a comment that says why this just returns true? Is it something that is a TODO, is it something that should be added by the consumer of this class? Something we were initially checking client-side but switched to server-side and it's just for maintaining existing code?

File: packages/iris-grid/src/formatters/TableColumnFormatter.ts [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+  static isSameFormat(
+    formatA?: TableColumnFormat,
+    formatB?: TableColumnFormat
+  ): boolean {
+    throw new Error('isSameFormat not implemented');
+  }
+
+  /**
+   * Returns format object
+   */
```
Comment (mattrunyon) [inline packages/iris-grid/src/formatters/TableColumnFormatter.ts:50]: Add param/return tags

File: packages/iris-grid/src/formatters/DateTimeColumnFormatter.ts [hunk 24]
```diff
@@ -128,11 +157,11 @@ class DateTimeColumnFormatter extends TableColumnFormatter {
     );
   }
 
-  getEffectiveFormatString(baseFormatString) {
+  getEffectiveFormatString(baseFormatString: string): string {
     return this.formatStringMap.get(baseFormatString) || baseFormatString;
   }
 
-  format(value, format) {
+  format(value: unknown, format?: TableColumnFormat): string {
```
Comment (mattrunyon) [inline packages/iris-grid/src/formatters/DateTimeColumnFormatter.ts:164]: Can this value be set using the generic in `TableColumnFormatter`? Or is it always unknown? If always unknown, the `TableColumnFormatter` class might not need to have a generic type

File: packages/iris-grid/src/formatters/TableColumnFormatter.ts (hunk 22 above)
Comment (mattrunyon) [reply packages/iris-grid/src/formatters/TableColumnFormatter.ts:33]: Nevermind I saw the comment at the top of the file saying it's the default formatter

Should we maybe use `abstract` classes and methods in this case? At least the class should be `abstract` I think if it's supposed to be extended and not directly used. Not sure if we want defaults for the methods or mark some `abstract` so the implementation has to define it

**Review by mattrunyon — CHANGES_REQUESTED** [review]

File: packages/iris-grid/src/formatters/TableColumnFormatter.ts (hunk 22 above)
Comment (mofojed) [reply packages/iris-grid/src/formatters/TableColumnFormatter.ts:33]: Yea it _should_ be an abstract class, or possibly even just an interface. I added a comment on #21 to consider that change. I didn't want to change any of the existing functionality as I was just changing the one class so I don't get errors in Linker

File: packages/iris-grid/src/formatters/DateTimeColumnFormatter.ts (hunk 24 above)
Comment (mofojed) [reply packages/iris-grid/src/formatters/DateTimeColumnFormatter.ts:164]: I think always `unknown`. I will revisit when converting grid to TS (#21)

File: packages/iris-grid/src/formatters/TableColumnFormatter.ts (hunk 21 above)
Comment (mofojed) [reply packages/iris-grid/src/formatters/TableColumnFormatter.ts:2]: I think it might've been for params when it was a js file - for some reason it complains about unused params in JS, in TS it does not. I've removed since this file is now ts.

**Review by mattrunyon — APPROVED** [review]
